organizations/plugins.go

```diff
@@ -1,269 +1,346 @@
-		// plugin not found in organization.
-		utils.GetError(errors.New("plugin does not exist"), http.StatusBadRequest, w)
-		return
-	}
-
-	orgPluginObjID := orgPluginDoc["_id"]
-	orgPluginID := orgPluginObjID.(primitive.ObjectID).Hex()
-
-	doc, err := utils.DeleteOneMongoDBDoc(orgCollectionName, orgPluginID)
-
-	if err != nil {
-		// plugin not found in organization.
-		utils.GetError(err, http.StatusNotFound, w)
-		return
-	}
-
-	if doc.DeletedCount == 0 {
-		utils.GetError(errors.New("plugin failed to uninstall"), http.StatusBadRequest, w)
-		return
-	}
-
-	utils.GetSuccess("plugin removed successfully", doc, w)
-}
+package organizations
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"zuri.chat/zccore/utils"
+)
+
+var ErrNoAuthToken = errors.New("no authorization header provided")
+
+// Install plugin into an organization.
+func (oh *OrganizationHandler) AddOrganizationPlugin(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	var orgPlugin OrgPluginBody
+
+	OrgID := mux.Vars(r)["id"]
+
+	err := json.NewDecoder(r.Body).Decode(&orgPlugin)
+	if err != nil {
+		utils.GetError(err, http.StatusBadRequest, w)
+		return
+	}
+
+	// confirm if plugin_id exists
+	pluginID, err := primitive.ObjectIDFromHex(orgPlugin.PluginID)
+
+	if err != nil {
+		utils.GetError(errors.New("invalid plugin id"), http.StatusBadRequest, w)
+		return
+	}
+
+	plugin, _ := utils.GetMongoDBDoc(PluginCollectionName, bson.M{"_id": pluginID})
+
+	if plugin == nil {
+		utils.GetError(errors.New("operation failed"), http.StatusBadRequest, w)
+		return
+	}
+
+	// confirm if user_id exists
+	creatorID, err := primitive.ObjectIDFromHex(orgPlugin.UserID)
+
+	if err != nil {
+		utils.GetError(errors.New("invalid user id"), http.StatusBadRequest, w)
+		return
+	}
+
+	user, _ := utils.GetMongoDBDoc(MemberCollectionName, bson.M{"_id": creatorID, "org_id": OrgID})
+	if user == nil {
+		utils.GetError(errors.New("member doesn't exist in the organization"), http.StatusBadRequest, w)
+		return
+	}
+
+	var member Member
+	if err = utils.ConvertStructure(user, &member); err != nil {
+		utils.GetError(err, http.StatusInternalServerError, w)
+		return
+	}
+
+	if member.Role != OwnerRole && member.Role != AdminRole {
+		utils.GetError(errors.New("access denied"), http.StatusForbidden, w)
+		return
+	}
+
+	orgID, err := primitive.ObjectIDFromHex(OrgID)
+
+	if err != nil {
+		utils.GetError(errors.New("invalid organization id"), http.StatusBadRequest, w)
+		return
+	}
+
+	p, _ := utils.GetMongoDBDoc(OrganizationCollectionName, bson.M{"_id": orgID},
+		options.FindOne().SetProjection(bson.D{{Key: PluginCollectionName, Value: 1}, {Key: "_id", Value: 0}}))
+
+	plugins := make([]map[string]interface{}, 0)

```

Comment:
I was thinking, instead of a a list of maps, why not a map. That way we can get rid of the for loops and operate at constant time when checking if a plugin exists or not. What do you think?